Feature/question1

big-Miyabi opened the pull request:
課題
---
- [課題1 · Issue #1 · Caraquri/intern_2week_study_ios](https://github.com/Caraquri/intern_2week_study_ios/issues/1])

概要
---
- `addButton`を押した際に`textField`に入っている文字を`textView`に追加する処理を実装
  - 起動時に`textView`を空にする
  - `textView`は改行してから追加する
- `clearTextButton`を追加し、`textView`の文字を消す処理を実装
- `addButton`を押した際、`textField`に文字が入っていない場合は`textView`に追加できないようにする処理を追加
  - その際は、`TextField`の下にLabelを追加し「赤文字」で「文字を入力してください」と警告を表示

特に確認して頂きたい項目
---
- ボタン押下時のアクションの書き方が適切か
  - `@IBAction`を使う方法と`UIButton.addTarget`を使う方法の二つを見つけたが、どちらを使用すべきか
  - 今回は、どのボタンのアクションなのか一眼見て分かる後者が適切だと思ったため、`UIButton.addTarget`を使用した

- `addFuncToButton`関数が適切か
  - `addTarget`を使う際、
`addTextButton.addTarget(self, action: #selector(addText(_:)),for: UIControl.Event.touchUpInside)`という風に書いていたが、これだと処理が長い上にパッと見たときに何をしているのかすぐに分かりづらいかなと思ったため、`addFuncToButton`関数を作成して簡潔に処理を書けるようにした
  - 自分は見やすいと思い作成しましたが、これについてどう思うかお聞きしたいです

- 他にもっと良い処理の実装方法はあるか

動作確認用 Screenshots
---
<img src="https://user-images.githubusercontent.com/47961006/89991933-86674300-dcbf-11ea-9216-e9c0dd29788e.gif" width="300" />

参考URL
---
[【Swift】addTargetメソッドの使い方、引数selector、senderのおさらい](https://pg-happy.jp/swift-addtarget.html)

big-Miyabi committed:
Add: 課題1-2までクリア

Add
Addボタン押下時にtextFieldに入っている文字をtextViewに追加する処理を実装
Clearボタン押下時にtextViewの文字を消す処理を実装

big-Miyabi committed:
Add: 課題1-3をクリア(Warningの表示)

big-Miyabi committed:
Add: addFuncToButton関数を追加

Add:
今まではaddTextButton.addTarget(self, action: #selector(addText(_:)),for: UIControl.Event.touchUpInside)という風に書いていたが、
これだと処理が長い上にパッと見たときに何をしているのかすぐに分かりづらいかなと思ったので、
addFuncToButton関数を作成して簡潔に処理を書けるようにした

big-Miyabi committed:
Rename: addTextButton -> buttonToAddText, clearTextButton -> buttonToClearText

Rename:
addTextButtonとclearTextButtonは、動詞が先頭に来ているため関数名と勘違いする可能性があるので、buttonToAddTextというように明確な名前に変更した

特に、前回addFuncToButton関数を作成した影響で、
addFuncToButton(addTextButton, #selector(addText(_:)))
というような、全てaddから始まる非常にややこしいコードになっていたので、早急に変える必要があった。

big-Miyabi committed:
Move: viewDidLoad()が一番上に来るように関数の順番を変えた

snabezawa-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -3,10 +3,33 @@ import UIKit
     @IBOutlet weak var textField: UITextField!
-    @IBOutlet weak var addTextButton: UIButton!
+    @IBOutlet weak var warningLabel: UILabel!
+    @IBOutlet weak var buttonToAddText: UIButton!
+    @IBOutlet weak var buttonToClearText: UIButton!
```
変数名がおかしいかも🤔
UI ComponentはxxxButtonやxxxLabelというように末尾に型の名称を入れるとSwiftっぽくかけます！
なので`addButton`, `clearButton`のほうがわかりやすいです。

big-Miyabi committed:
Remove: アンラップする必要がない型に!を付けていたので消した

snabezawa-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -3,10 +3,33 @@ import UIKit
+        warningLabel.isHidden = true
+    }
+    
+    func addFuncToButton(_ button: UIButton!, _ selector: Selector) {
+        button.addTarget(self, action: selector, for: UIControl.Event.touchUpInside)
+    }
```
addTargetを追加する関数は実務で書かないかも〜
addFuncToButton~は役割が少ないので、メソッドで切り分けなくてもいいかもです。

それよりかはbuttonのスコープ内で書くかも。
以下のような感じでプロパティに初期値を与えることができます。


```
@IBOutlet weak var addButton: UIButton! {
    didSet {
        addButton.addTarget(self, action: #selector(addText(_:), for: UIControl.Event.touchUpInside)
    }
}
```

snabezawa-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -3,10 +3,33 @@ import UIKit
         super.viewDidLoad()
+        addFuncToButton(buttonToAddText, #selector(addText(_:)))
+        addFuncToButton(buttonToClearText, #selector(clearText(_:)))
+        textView.text = ""
```
空文字を入れてるのはなんででしょう？🤔 
画面がロードされる時は空文字では。。？

snabezawa-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -3,10 +3,33 @@ import UIKit
+    }
+    
+    @objc func addText(_ sender: UIButton) {
+        if textField.text! == "" {
```
強制アンラップNG🙅‍♀️
guard文でoptionalを外してください。
あと空文字判定は `isEmpty`があります。

snabezawa-cq commented on intern_2week_study/Question/Question1/Question1ViewController.swift:
```diff
@@ -3,10 +3,33 @@ import UIKit
+            return
+        }
+        warningLabel.isHidden = true
+        textView.text += "\n" + textField.text!
```
ここも `!`

big-Miyabi committed:
Change: buttonToAddText -> addButton,  buttonToClearText -> clearButton
Remove: storyboard上のUILabel内の初期値を空文字に変更し、それに伴いviewDidLoad()時のtextView.text = ""を削除した

big-Miyabi committed:
Improve: addFuncToButton関数を削除し、UIButtonのスコープ内にaddTargetの処理を加えた

big-Miyabi committed:
Fix: 強制アンラップはせずに、guard letとisEmptyを用いて処理を書き換えた
